### DISCUSSION TITLE
Communication between glsp-theia and glsp-client

### DISCUSSION THREAD

**Discussion by wss29** [eclipse-glsp/glsp#942] [Q&A] [answered]
Hi,
    you know there are three parts for a glsp application in theia. 
v1-glsp,
v1-server,
v1-theia. 
Now I create a custom widget in v1-theia, and I add a button in my widget. when I click the button I want to send something to v1-glsp. Currently I can get opened GLSPDiagramWidget's actionDispatcher, but the dispatcher can only send messages to v1-server,  I do not need to deal with things in v1-server. I read the document https://www.eclipse.org/glsp/documentation/actionhandler/#overview. It says that the client can handle action, but I do not find a good example to do that. All I see is handle action in the server. could anyone give some advice?

**Comment by tortmayr** [comment, marked as answer]
Hi @wss29 ,
>Currently I can get opened GLSPDiagramWidget's actionDispatcher, but the dispatcher can only send messages to v1-server, 

That's not correct, the `ActionDispatcher` of the `GLSPDiagramWidget` can dispatch actions to both the server and the client.
All you have to do is to configure an `ActionHandler` for your action in your `v1-glsp`. The GLSP documentation has a dediacted section on how to configure client action handlers: https://www.eclipse.org/glsp/documentation/actionhandler/#implementing-an-action-handler-glsp-client

**Reply by wss29** [reply]
Hi,
    thanks, tons of code recently, I forget the document detail. It is ok now
